### ISSUE TITLE
WinGUI beep() causing program to pause

### ISSUE THREAD

**Issue by rmorales87atx** [Bill-Gray/PDCursesMod#190]
I recently updated to PDCursesMod 4.2.0 and noticed an obscure issue where my program appeared to be pausing at random. The pauses never happened on an X11 build. I finally paid attention to notice the "pause" was happening every time I had `beep()` called.

An inspection of `PDC_beep` in `wingui\pdcutil.c` suggests that [`PlaySound`](https://docs.microsoft.com/en-us/previous-versions/dd743680(v=vs.85)) is being called without the `SND_ASYNC` flag, and thus causing a very noticeable delay; my call to `beep()` does not return for nearly 3-4 seconds! **Adding `SND_ASYNC` and recompiling resolves the issue.**

Suggested edit in `wingui\pdcutil.c`:
```
void PDC_beep(void)
{
    PDC_LOG(("PDC_beep() - called\n"));

    LeaveCriticalSection(&PDC_cs);
    if (!PlaySound((LPCTSTR) SND_ALIAS_SYSTEMDEFAULT, NULL, SND_ALIAS_ID|SND_ASYNC))
        Beep(800, 200);
    EnterCriticalSection(&PDC_cs);
}
```

It appears that WinCon uses the same call to `PlaySound` in `wincon\pdcutil.c`, and thus the same fix of adding the `SND_ASYNC` flag is suggested.

**Comment by rmorales87atx** [comment]
Just now reading through the commit history in `pdcutil.c`, it seems the async flag was deliberately left off to allow "multiple" beeps to succeed. I'm guessing the situation I ran into wasn't in your test. My program is basically issuing a `beep()`, then it does some screen updating and a pop-up dialog appears; the pop-up does not appear until 3+ seconds after the `beep()`.

For the multiple beep issue, maybe we can try setting the flags to `SND_ALIAS_ID|SND_ASYNC|SND_NOSTOP`?
```
SND_NOSTOP        If this flag is not specified, PlaySound attempts to stop any sound that is currently playing in the same process. Sounds played in other processes are not affected.
```

**Comment by GitMensch** [comment]
Whoa, I'd call that a severe issue with 4.2.0. For reference: the issue solved with that was #176, included with #179.

@rmorales87atx Did you already found the time to check your suggested fix?
@loblolly986 I think this approach would still work for you, wouldn't it? After all the plan was to "keep the window responsive".

i _think_ using `SND_NOSTOP` _may_ be the wrong thing - I'm not sure what happens if you issue a beep every second (and the beep itself takes more than one) - is the user experience better with multiple aborted/partial beeps or with multiple "overlapped" ones?

**Comment by rmorales87atx** [comment]
> Whoa, I'd call that a severe issue with 4.2.0. For reference: the issue solved with that was #176, included with #179.
> 
> @rmorales87atx Did you already found the time to check your suggested fix?

I tested with `SND_ASYNC` before posting this issue and that did resolve the problem for me. I have not yet tested with `SND_NOSTOP`, might be able to do that today.

**Comment by rmorales87atx** [comment]
Right now with just `SND_ASYNC` added, if I trigger my program enough times I get multiple beeps that do sort of "cut" each other - but the last beep does finish all the way; IMO, this "sounds right" to me based on how other Windows programs behave.

OK, now I agree with `SND_NOSTOP` being wrong. Testing with that flag, triggering multiple beeps appears to cause `PlaySound` to fail and thus the fallback to `Beep(800, 200)` is triggered; it's a bit more "odd" hearing `Beep()` in between what `PlaySound` actually triggers.

**Comment by rmorales87atx** [comment]
Here's a few more datapoints:

- Dragged out a Windows 7 VM and confirmed the async problem.
- On Windows 7 multiple async beeps are a bit more severely "cut" than they are on current Windows 10.
- Commenting out the `PlaySound` call and using Win32 `Beep()` still results in a 200ms delay (though much more "tolerable" than 3 seconds!)
- However: rapid-fire `Beep()` calls in quick succession still result in a "choppy" sound

IMO, I don't think this "problem" with multiple beeps can be solved in a realistic manner. I also can't think of a realistic situation where a program would need multiple beeps in such quick succession; I'd think at that point the programmer would want more direct control than what `beep()` offers anyway.

**Comment by GitMensch** [comment]
I'd _guess_ that the async is no problem (because it is programmer-done when it occurs) but the solution to the "pause issue" while keeping the original solution which was named "beeps to play async" ;-)

**Comment by Bill-Gray** [comment]
I've tried my fix a6212d94d26a33ea9c93d6ba12b1e7a0e0c5e5ea on Wine and Win10,  for both WinCon and WinGUI,  and it appears to work properly,  by having the beeps handled in a separate thread,  with some logic to count the beeps.  Run `tuidemo` and click on `TwoBeeps` twice,  and the program returns immediately (you see "Two Beeps! Two Beeps!" on the screen).  The four beeps are then played in sequence in the other thread,  while the main thread carries on normally.

I'd thought the multithreading might be a complicated mess,  but it's actually reasonably straightforward (essentially the same on both platforms).

The pause between beeps is longer than I think you'd really want.  It could be that something other than `SND_ALIAS_SYSTEMDEFAULT` should be used... but that's really a separate issue,  as is the lack of beeps on VT and SDLn.

It occurs to me that the new `PDC_beep()` code is identical now in WinCon and WinGUI.  It could be moved to the `common` directory,  and then also used in Windows builds for VT and SDLn.

**Comment by rmorales87atx** [comment]
> The pause between beeps is longer than I think you'd really want. It could be that something other than `SND_ALIAS_SYSTEMDEFAULT` should be used... but that's really a separate issue, as is the lack of beeps on VT and SDLn.

I have to admit I'm really not fond of this change... if I bang up my program a bit to make it beep a lot, I could be listening to beeps for well over a few minutes. But, as you say it's now a separate issue. *(For now I'll just change my program to use `MessageBeep()` and not PDC's `beep()`.)*

**Comment by Bill-Gray** [comment]
I'm having some second thoughts about this,  too.  I could see potential problems with this if other Curses implementations don't do the same thing.

 I can't get ncurses' `beep()` to work at all.  On that platform, `tuidemo`'s Two Beeps returns right away.  I don't know if ten beeps in a row would be stacked up if the noise worked,  or if you'd just get one beep (as happens with @wmcbrine's version,  and would have happened here before we started playing around with `beep()`.)

If "combining" beeps is the norm,  I could imagine somebody writing code,  casually beeping on a slew of errors and expecting only one beep to come out.  Or never even noticing that their code was sending a dozen beeps unintentionally,  because all they ever heard was one beep.  In that case,  they could be _really_ annoyed to suddenly get a barrage of beeps when they used PDCursesMod.

If so,  I think we'd decide to do a lot of reversion until we reach the previous state,  in which if you want two beeps,  you have to beep,  wait a bit,  then beep again.
